internal/internal_workflow_client.go

```diff
@@ -1170,3 +950,273 @@ func serializeSearchAttributes(input map[string]interface{}) (*commonpb.SearchAt
+	taskTimeout := in.Options.WorkflowTaskTimeout
+
+	// Encode input
+	input, err := encodeArgs(dataConverter, in.Args)
+	if err != nil {
+		return nil, err
+	}
+
+	memo, err := getWorkflowMemo(in.Options.Memo, dataConverter)
+	if err != nil {
+		return nil, err
+	}
+
+	searchAttr, err := serializeSearchAttributes(in.Options.SearchAttributes)
+	if err != nil {
+		return nil, err
+	}
+
+	// get workflow headers from the context
+	header, err := headerPropagated(ctx, w.client.contextPropagators)
+	if err != nil {
+		return nil, err
+	}
+
+	signalWithStartRequest := &workflowservice.SignalWithStartWorkflowExecutionRequest{
+		Namespace:                w.client.namespace,
+		RequestId:                uuid.New(),
+		WorkflowId:               in.Options.ID,
+		WorkflowType:             &commonpb.WorkflowType{Name: in.WorkflowType},
+		TaskQueue:                &taskqueuepb.TaskQueue{Name: in.Options.TaskQueue, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		Input:                    input,
+		WorkflowExecutionTimeout: &executionTimeout,
+		WorkflowRunTimeout:       &runTimeout,
+		WorkflowTaskTimeout:      &taskTimeout,
+		SignalName:               in.SignalName,
+		SignalInput:              signalInput,
+		Identity:                 w.client.identity,
+		RetryPolicy:              convertToPBRetryPolicy(in.Options.RetryPolicy),
+		CronSchedule:             in.Options.CronSchedule,
+		Memo:                     memo,
+		SearchAttributes:         searchAttr,
+		WorkflowIdReusePolicy:    in.Options.WorkflowIDReusePolicy,
+		Header:                   header,
+	}
+
+	var response *workflowservice.SignalWithStartWorkflowExecutionResponse
+
+	// Start creating workflow request.
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+
+	response, err = w.client.workflowService.SignalWithStartWorkflowExecution(grpcCtx, signalWithStartRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	iterFn := func(fnCtx context.Context, fnRunID string) HistoryEventIterator {
+		rpcScope := metrics.GetMetricsScopeForRPC(w.client.metricsScope, in.WorkflowType,
+			metrics.NoneTagValue, in.Options.TaskQueue)
+		return w.client.getWorkflowHistory(fnCtx, in.Options.ID, fnRunID, true,
+			enumspb.HISTORY_EVENT_FILTER_TYPE_CLOSE_EVENT, rpcScope)
+	}
+
+	curRunIDCell := util.PopulatedOnceCell(response.GetRunId())
+	return &workflowRunImpl{
+		workflowType:  in.WorkflowType,
+		workflowID:    in.Options.ID,
+		firstRunID:    response.GetRunId(),
+		currentRunID:  &curRunIDCell,
+		iterFn:        iterFn,
+		dataConverter: w.client.dataConverter,
+		registry:      w.client.registry,
+	}, nil
+}
+
+func (w *workflowClientInterceptor) CancelWorkflow(ctx context.Context, in *ClientCancelWorkflowInput) error {
+	request := &workflowservice.RequestCancelWorkflowExecutionRequest{
+		Namespace: w.client.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: in.WorkflowID,
+			RunId:      in.RunID,
+		},
+		Identity: w.client.identity,
```

Comment:
and the RequestId here as well.

Reply:
Also commented at https://github.com/temporalio/sdk-go/pull/610#discussion_r746670411. I think we need a new issue for this instead of inside this PR.